llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeFlashAttentionPass.cpp

```diff
@@ -0,0 +1,420 @@
+
+    // Update global statistics
+    Value newMax = computeNewMax(iterArgMax, currentMax, empty, loc, rewriter);
+    Value oldWeight =
+        subtractAndExponentiate1D(iterArgMax, newMax, empty, loc, rewriter);
+    Value currentWeight =
+        subtractAndExponentiate1D(currentMax, newMax, empty, loc, rewriter);
+    Value newSum = computeNewSum(iterArgSum, currentSum, oldWeight,
+                                 currentWeight, empty, loc, rewriter);
+
+    // Compute softmax
+    Value softmax = computeSoftmax(qkTranspose, currentMax, currentWeight,
+                                   newSum, emptySquare, loc, rewriter);
+
+    // Update accumulator
+    empty = rewriter.create<tensor::EmptyOp>(
+        loc, SmallVector<OpFoldResult>{sequenceLength, headDimension},
+        elementType);
+    Value scaledAcc = scaleAccumulator(outputSlice, iterArgSum, newSum,
+                                       oldWeight, empty, loc, rewriter);
+
+    // Compute matmul(softmax, v)
+    Value result = rewriter
+                       .create<linalg::MatmulOp>(
+                           loc, outputSlice.getType(),
+                           ValueRange{softmax, valueSlice}, scaledAcc)
+                       .getResult(0);
+
+    // Insert slices
+    auto [updatedAcc, updatedMax, updatedSum] =
+        insertSlices(result, iterArgResult, newMax, max, newSum, sum,
+                     queryShape, ivs, sequenceTileLength, loc, rewriter);
+
+    if (scf::YieldOp yieldOp = dyn_cast<scf::YieldOp>(
+            secondLoopNest.loops.back().getBody()->getTerminator())) {
+      rewriter.replaceOpWithNewOp<scf::YieldOp>(
+          yieldOp, ValueRange{updatedAcc, updatedMax, updatedSum});
+    }
+
+    if (scf::YieldOp yieldOp = dyn_cast<scf::YieldOp>(
+            firstLoopNest.loops.back().getBody()->getTerminator())) {
+      rewriter.setInsertionPoint(yieldOp);
+      rewriter.replaceOpWithNewOp<scf::YieldOp>(
+          yieldOp, ValueRange{secondLoopNest.results[0]});
+    }
+
+    attnOp.getResults()[0].replaceAllUsesWith(firstLoopNest.results[0]);
+    return success();
+  }
+};
+
+} // namespace
+
+namespace {
+struct TileAndDecomposeFlashAttentionTransformPass
+    : public TileAndDecomposeFlashAttentionTransformBase<
+          TileAndDecomposeFlashAttentionTransformPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AffineDialect, IREE::LinalgExt::IREELinalgExtDialect,
+                    linalg::LinalgDialect, scf::SCFDialect,
+                    tensor::TensorDialect>();
+  }
+
+  void runOnOperation() override;
+};
+} // namespace
+
+void TileAndDecomposeFlashAttentionTransformPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  RewritePatternSet patterns(&getContext());
+  patterns.insert<ReifyFlashAttentionFwdTransform>(context);
+  if (failed(
+          applyPatternsAndFoldGreedily(getOperation(), std::move(patterns)))) {
+    return signalPassFailure();
```

Comment:
Not for this patch to solve, but I really wonder if the complexity of the greedy pattern rewriter is warranted just to lower one specific op with no pattern matching involved. Sounds like a simple walk over the IR with an `IRRewriter` would be largely sufficient.

Reply:
+1 to that. Its a habit in MLIR, that we (including me) have to unlearn.

Reply:
Same here. It is one of those handy tools that you habitually gravitate towards. Will change.